src/utils/communication_buffer.hpp

```diff
@@ -0,0 +1,409 @@
+//========================================================================================
+// (C) (or copyright) 2020-2022. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+
+#ifndef UTILS_COMMUNICATION_BUFFER_HPP_
+#define UTILS_COMMUNICATION_BUFFER_HPP_
+
+#include <type_traits>
+#include <iostream>
+#include <memory>
+#include <utility>
+#include <unordered_map>
+
+#ifdef MPI_PARALLEL
+#include <mpi.h>
+
+#define request_t MPI_Request
+#define comm_t MPI_Comm
+#else
+#define request_t int
+#define comm_t int
+#endif
+
+namespace parthenon { 
+namespace detail
+{
+
+#ifdef MPI_PARALLEL
+  // MPIType<A> -> MPIType<A, bool_t<true>> and then
+  // specialization checks are performed. This is why the
+  // default parameter is required.
+  template <bool>
+  struct bool_t
+  {
+  };
+
+  template <class T, class = bool_t<true>>
+  struct MPIType;
+
+  template <class U, class T>
+  using MPI_type_check = std::is_same<U, std::remove_pointer_t<T>>;
+
+  template <class T>
+  struct MPIType<T, bool_t<MPI_type_check<double, T>::value>>
+  {
+    MPI_Datatype static value() noexcept { return MPI_DOUBLE; }
+  };
+
+  template <class T>
+  struct MPIType<T, bool_t<MPI_type_check<Real, T>::value>>
+  {
+    MPI_Datatype static value() noexcept { return MPI_PARTHENON_REAL; }
+  };
+
+  template <class T>
+  struct MPIType<T, bool_t<MPI_type_check<int, T>::value>>
+  {
+    MPI_Datatype static value() noexcept { return MPI_INT; }
+  };
+
+  template <class T>
+  struct MPIType<T, bool_t<MPI_type_check<bool, T>::value>>
+  {
+    MPI_Datatype static value() noexcept { return MPI_CXX_BOOL; }
+  };
```

Comment:
is this more generic than comms buffers? I.e., should there be an mpi header where this lives? It mimics to some degree the reductions stuff that @jdolence added a while back

Reply:
Yeah, I wrote this for the test code and didn't notice that there was something similar already in Parthenon. I think this is a little more generic, but it probably easier to just to switch to the preexisting version.

Reply:
switched.